R/tmb-methods.R

```diff
@@ -429,3 +429,152 @@ h_residuals_response <- function(object) {
   assert_class(object, "mmrm_tmb")
   component(object, "y_vector") - unname(fitted(object))
 }
+
+
+#' @describeIn mmrm_tmb_methods Simulates from an mmrm model object
+#' @importFrom stats simulate
+#' @exportS3Method
+#'
+#' @param object an object of class 'mmrm'
+#' @param nsim number of response vectors to simulate. Defaults to 1.
+#' @param seed an object specifying if and how the random number generator should be initialized (‘seeded’).
+#' @param newdata data frame for which to evaluate predictions.
+#' @param ...
+#'
+#' @return A \code{data.frame} of dimension [n, m] where n is the number of observations
+#' fitted in the model, and m is the number \code{nsim} of simulated responses.
+simulate.mmrm_tmb <- function(object, nsim = 1,
+                              newdata = NULL,
+                              method = "conditional",
+                              ...){
+
+  method <- match.arg(method)
+
+  if(is.null(newdata)){
+    # build data.frame of prediction vectors
+    ret <- fitted(object)
+    ret <- data.frame(replicate(nsim, ret))
+
+    # get covariance matrix
+    covmat <- VarCorr(object)
+    grouped <- if(object$tmb_data$n_groups > 1) TRUE else FALSE
+
+    # get dataframe values
+    n_subjects <- object$tmb_data$n_subjects
+    subject_n_visits <- object$tmb_data$subject_n_visits
+    subject_zero_inds <- object$tmb_data$subject_zero_inds
+    visit_zero_inds <- object$tmb_data$visits_zero_inds
+
+    for(i in 1:n_subjects){
+
+      # Obtain indices of data.frame belonging to subject i
+      num_visits <- subject_n_visits[i]
+      inds <- (subject_zero_inds[i] + 1) : (subject_zero_inds[i] + num_visits)
+      visits <- visits_zero_inds[inds] + 1
+
+      # get relevant covariance matrix for subject i
+      covmat_i <- if(grouped){
+        covmat[[object$tmb_data$subject_groups[i]]]
+      }else{
+        covmat
+      }
+      # subset covariance matrix
+      covmat_i <- covmat_i[visits, visits]
+
+      # simulate from covariance matrix
+      ret[inds,] <- ret[inds, , drop=FALSE] + MASS::mvrnorm(nsim, rep.int(0, num_visits), covmat_i)
+    }
+  }else if(method == "conditional"){
+    # order data.frame
+    newdata <- dplyr::arrange(newdata, !!object$formula_parts$subject_var, !!object$formula_parts$visit_var)
```

Comment:
reordering can cause issues as the input and output are not corresponding to each other

Reply:
and this reordering can be wrong; suggest to use `model.frame.mmrm_tmb` to obtain the correct data

Reply:
e.g. data and fit do not have identical levels